### PR TITLE
[DS-3673] First try:  get latest COUNTER robots list release

### PR DESCRIPTION
**Pull request by mwoodiupui**
## References
Fixes #7020

## Description
Command line gadget to get a plain-text list of robot agent patterns from the latest release by COUNTER.  This is not a *complete* response to DS-3673 but it is a start.

## Instructions for Reviewers
New feature.

List of changes in this PR:
* New class `org.dspace.app.statistics.GetGithubRelease`, a general-purpose gadget to fetch the latest release of any Github project and either save the Zip archive as-is or unpack one or more files from it.
* Launcher configuration specific to the COUNTER-Robots project.

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.  (Except `main()`, because we all know what `main()` does.)
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Comment by tdonohue** [comment]
@mwoodiupui : If you are willing to use the new Scripts & Processes framework (instead of the old `launcher.xml`), then you could run the script directly in Integration tests using this `runDSpaceScript()` method:
https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/test/java/org/dspace/app/rest/test/AbstractIntegrationTestWithDatabase.java#L232

...and you could write ITs at the REST API layer (as any script of this type is runnable either via the REST API or via CLI...whereas `launcher.xml` is CLI only).  

We've only moved a handful of scripts from `launcher.xml` to Scripts & Processes so far, but they are all listed here: https://github.com/DSpace/DSpace/blob/main/dspace/config/spring/api/scripts.xml

So, you could use those as examples.   If you look closely, the `index-discovery` script no longer appears in `launcher.xml`, but it *will* respond on the commandline as normal (`./dspace index-discovery`)...or you can now run it from the Admin UI / REST API (see the example "Processes Admin UI" video in the beta 3 announcement; https://lyrasisnow.org/now-available-dspace-7-0-beta-3/)

**Comment by mwoodiupui** [comment]
I've turned it into a "script" and added a bit of testing.  But I wonder whether this really should be accessible from the UI.  It's really meant to be run regularly by the system scheduler to keep a robot list updated.  Its only function is to grab a release from Github and stash it on the server's filesystem.  Shouldn't this really be command-line-only?

Converting it was instructive.  Now I need to find out how a "script" is supposed to return errors or status.

**Comment by tdonohue** [comment]
Per decision in DSpace Steering (and as announced in last DSpace DevMtg), I'm moving this to the 8.0 board. Steering has limited 7.6 to only features which provide "feature parity/compatibility" with 6.x. Even though this will move to the 8.0 board, our goal is to move all existing pre-8.0 features forward as soon as 7.6 development is finishing (as 7.6 will be the last 7.x release allowing new features).

File: dspace-api/src/main/java/org/dspace/app/statistics/GetGithubRelease.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            return;
+        }
+
+        // Not help.  Remember whether verbose was required.
+        verbose = commandLine.hasOption(GetGithubReleaseOptions.OPT_VERBOSE);
+
+        // Identify the latest release.
+        String owner = commandLine.getOptionValue(GetGithubReleaseOptions.OPT_OWNER);
+        String repo = commandLine.getOptionValue(GetGithubReleaseOptions.OPT_REPO);
+        URL releaseConnection = new URL(String.format("https://api.github.com/repos/%s/%s/releases/latest",
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/statistics/GetGithubRelease.java:64]: Do we want to blindly always gather the latest version?  Or should we really just be taking in a URL here as a parameter... that way it's possible to use a specific tagged version of the files (which could be useful if one version of DSpace supports a different file format than another)

If you feel this is the best approach, I'm OK with it.  I just wanted to point out that blindly taking the "latest" can mean that you might try to download an update for DSpace 7 only to find that the new "latest" version is only compatible with DSpace 8 (as an example)... I don't know if this would ever happen, but wanted to point it out.

File: dspace/config/launcher.xml [hunk 2]
```diff
@@ -96,6 +96,19 @@
             <class>org.dspace.app.sitemap.GenerateSitemaps</class>
         </step>
     </command>
+    <command>
+        <name>get-counter-robots</name>
```
Comment (tdonohue) [inline dspace/config/launcher.xml:100]: Should this be named "update-counter-robots"?

I also wonder if this should install the file in the correct location?  Personally, I wouldn't know where to place this file.  Does it go under `[dspace]/spiders/` or `[dspace]/spiders/agents`?

If it doesn't put the file in the correct location, then we need to provide clear documentation on how to install this file.  Ideally, though, if we want people to be able to run this via a cron job, it'd be nice to have it install the file in the proper place.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@mwoodiupui : Apologies for the very long delay in feedback on this PR.  I'm just revisiting it today, and I'm finding that I cannot get this tool to work (or maybe don't know how to use it?).

I'm trying this from commandline:
`./dspace get-counter-robots`

And I get this error:
```
Exception in thread "main" java.lang.NullPointerException
        at org.dspace.app.launcher.ScriptLauncher.runOneCommand(ScriptLauncher.java:293)
        at org.dspace.app.launcher.ScriptLauncher.handleScript(ScriptLauncher.java:134)
        at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:99)
```

I also agree with you that I don't think the "get-github-release" tool is useful for the Admin User interface.  It seems like an extremely advanced tool, that might only be used by other tools, but never **appear** even in the list of available tools (e.g. when running `./dspace` you get a list of every tool).  

In other words, I'm not sure it's useful for any use case other than this `get-counter-robots` script  at this time.  However, I do see the potential that other tools can use it if needed.

Beyond that, a few other minor notes inline below.

**Comment by alanorth** [comment]
I'm also wondering if this should always get the latest release, or if it should take a parameter specifying a release tag to download?

File: dspace/config/launcher.xml (hunk 2 above)
Comment (mwoodiupui) [reply dspace/config/launcher.xml:100]: It's called "get-" rather than "update-" because it doesn't update; it gets, and lets you decide whether you want to update.

It never occurred to me that someone would trust a cron job to do the updating.  YOU decide to run it because you have learned that there's a new release from COUNTER.  YOU inspect the file and decide whether it's suitable.  YOU install it.  This seems to be what was asked for in #7020.

Maybe we need to have the discussion that Bram suggested, before finishing this up?  People seem to have very different ideas of how to consume the COUNTER robots list.

Comment (tdonohue) [reply dspace/config/launcher.xml:100]: @mwoodiupui : Yes, it sounds like we may need to discuss the goals of this PR further.  

My **assumption** was that we'd want this to be as easy as possible to help sites keep this list _updated at all times._   So, I was assuming it'd be as simply as a scheduled cron job that you could run to update it one a week (if changed) or similar.

Your approach is much more "manual", which I'm worried means it's not much of an improvement over the current process.  An Admin could just go to this site and download the latest file quickly to inspect it & install it.  This script currently seems to only help them in the download process (which could also be done via a `wget` command).

So, I think you are correct that our two goals for this script don't currently align.  I'd rather have an update script here, and I'm questioning whether a simple "get" script is useful when we could just tell people to run something like this:
```
wget https://github.com/atmire/COUNTER-Robots/blob/master/generated/COUNTER_Robots_list.txt
```

**Comment by alanorth** [comment]
@mwoodiupui are you actively using COUNTER-Robots on your systems? The project has slowed down a lot and I've got a growing list of overrides in my local instance (with a few pull requests open upstream).

**Comment by tdonohue** [comment]
@mwoodiupui : stumbled back on this while reviewing all the `new feature` PRs on our 10.0 board.  If you feel this is still worth moving forward, please rebase this for review for 10.0.  However, I see @alanorth [asked whether this is worthwhile in a comment from 6 months ago](https://github.com/DSpace/DSpace/pull/2892#issuecomment-3114706806).

**Comment by tdonohue** [comment]
Unfortunately, this Feature PR missed our [DSpace 10.0 "Feature PR Merge Deadline"](https://wiki.lyrasis.org/display/DSPACE/DSpace+Release+10.0+Status#DSpaceRelease10.0Status-ReleaseTimeline). As DSpace 10.0 is now under a "Feature Freeze", this ticket will be moved to our 11.0 release board in the hopes of completing this feature in the next release.  If there are any questions, let us know.

**Comment by bram-atmire** [comment]
I had a look at the failing checks here (in the context of #12785). Both the Unit and Integration jobs fail on the same new class, `GetGithubReleaseIT`, and they share one root cause.

**1. Unit Tests job (Checkstyle)**

```
src/test/java/org/dspace/app/statistics/GetGithubReleaseIT.java:[19,8] (imports) UnusedImports: Unused import - org.junit.Ignore.
```

`org.junit.Ignore` is imported but never applied to a method.

**2. Integration Tests job**

`testInternalRunGetZip` errors on every rerun with:

```
java.lang.IllegalStateException: BeanFactory not initialized or already closed - call 'refresh' before accessing beans via the ApplicationContext
    at ...MockitoResetTestExecutionListener.beforeTestMethod(MockitoResetTestExecutionListener.java:97)
```

The error is raised by Spring's `MockitoResetTestExecutionListener` in `beforeTestMethod`, before your test body runs. It comes from mixing `@RunWith(SpringRunner.class)` + `@ContextConfiguration` on top of `AbstractIntegrationTest`. That base class (via `AbstractUnitTest`) builds and tears down the DSpace kernel itself with JUnit `@Before`/`@After` and is not designed to run under `SpringRunner`, so by the time the Spring `TestExecutionListener` tries to reset mocks the context has already been closed.

**The two failures are connected.** `GetGithubRelease` makes a live call to `https://api.github.com/repos/{owner}/{repo}/releases/latest`, so this test needs network access and can't run reliably in CI anyway. It looks like the intent was to `@Ignore` it (which is why the import is there). Applying `@Ignore` to the method fixes both jobs at once: the import becomes used (Checkstyle passes), and the skipped method never triggers `beforeTestMethod` (the IT error disappears).

Minimal fix:

```java
@Test
@Ignore("Contacts GitHub over the network to fetch a real release; run manually.")
public void testInternalRunGetZip()
        throws Exception {
```

If you also want to clean up the root cause so the class matches the other `AbstractIntegrationTest` subclasses, the `@RunWith(SpringRunner.class)` and `@ContextConfiguration(...)` annotations (and the `DSpaceKernelInitializer` / `DSpaceConfigurationInitializer` / `SpringRunner` / `RunWith` imports) aren't needed here. The test instantiates `GetGithubRelease` and `GetGithubReleaseScriptConfiguration` directly and only relies on the kernel and `eperson` that `AbstractIntegrationTest` already provides, so it can be a plain JUnit4 test:

```java
public class GetGithubReleaseIT
        extends AbstractIntegrationTest {
```

One related note for whenever this moves forward: the `atmire/COUNTER-Robots` list is now deprecated in favour of `Project-Counter/counter-bots` (see #12785), which you already have stubbed out in the commented `OWNER`/`REPO` constants. Heads up that #12843 removes the old opt-in Ant download of the Atmire list, so there shouldn't be any overlap with the launcher/scripts config here, but worth a rebase check.

Happy to help further if useful.
